File: hdf5/fortran/hdf5_3D.f90
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 ! Copyright by The HDF Group.                                                   !
 ! All rights reserved.                                                          !
 !                                                                               !
 ! The full HDF5 copyright notice, including terms governing use, modification,  !
 ! and redistribution, is contained in the file COPYING.  COPYING can be found   !
 ! at the root of the source code distribution tree.                             !
 ! For questions contact the help desk at help@hdfgroup.org                      !
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 !
 ! Programmer:  Isayah Reed
 !              May  4, 2011
 !
 ! Purpose:
 !  This is 1 of 3 simple HDF5 programs that demonstrates how to add CF attributes
 !  to an h5 file. This example shows how to use 3-D datasets. A file is created
 !  with 4 datasets: radiation, latitude, longitude, pressure. Radiation is a 3-D
 !  dataset that contains the CF attributes: units, fillvalue, long_name, and
 !  coordinates. Latitude, longitude, and pressure contain the CF attributes units
 !  and long_name. Data is written to 3D.h5
 !  
 ! Compile:
 !  $(HDF5_DIR)/bin/h5fc hdf5_3D.f90 -I$(HDF5_DIR)/include -L$(HDF5_DIR)/lib -lhdf5 -lhdf5_hl  -lhdf5hl_fortran -lhdf5_fortran -lz -lm
 !

PROGRAM main

  USE HDF5

  IMPLICIT NONE

  CHARACTER(LEN=6), PARAMETER :: filename = "3D.h5"
  CHARACTER(LEN=9) , PARAMETER :: RADIATION= "Radiation"
  CHARACTER(LEN=8) , PARAMETER :: LATITUDE= "latitude"
  CHARACTER(LEN=9) , PARAMETER :: LONGITUDE= "longitude"
  CHARACTER(LEN=8) , PARAMETER :: PRESSURE= "pressure"
  CHARACTER(LEN=9) , PARAMETER :: LONGNAME = "long_name"
  CHARACTER(LEN=5) , PARAMETER :: UNITS = "units"
  CHARACTER(LEN=10) , PARAMETER :: FILLVALUE = "_FillValue"
  CHARACTER(LEN=11) , PARAMETER :: COORDINATES = "coordinates"

  INTEGER          , PARAMETER :: nx1     = 180
  INTEGER          , PARAMETER :: ny1     = 360
  INTEGER          , PARAMETER :: nz1     = 20

  INTEGER :: hdferr
  INTEGER(HID_T) :: file, space, dset, attr_id, attr_space ! handles
  INTEGER(HID_T) :: atype_id
  CHARACTER(LEN=10), DIMENSION(3) ::  coorlist  ! Attribute string data
  ! switch dimensions to adhere to C standards
  INTEGER(HSIZE_T), DIMENSION(1:3) :: temp_dims = (/nz1,ny1,nx1/) ! size read/write buffer
  INTEGER(HSIZE_T), DIMENSION(1:1) :: lat_dims = (/nx1/) 
  INTEGER(HSIZE_T), DIMENSION(1:1) :: lon_dims = (/ny1/) 
  INTEGER(HSIZE_T), DIMENSION(1:1) :: press_dims = (/nz1/) 
  INTEGER(HSIZE_T), DIMENSION(1) :: dimsf ! dimension for scalars
  INTEGER(HSIZE_T), DIMENSION(2) :: dimsa ! dimension for arrays
  DOUBLE PRECISION, DIMENSION(1:nx1,1:ny1,1:nz1) :: temp_data
  DOUBLE PRECISION, DIMENSION(nx1) :: lat_data
  DOUBLE PRECISION, DIMENSION(ny1) :: lon_data
  DOUBLE PRECISION, DIMENSION(nz1) :: press_data
  DOUBLE PRECISION, DIMENSION(1) :: fill_data
  CHARACTER(LEN=30) :: attr_data
  INTEGER :: i, j, k, strlen


  !
  ! Initialize FORTRAN interface.
  !
  CALL h5open_f(hdferr)
  !
  ! Initialize temperature data.
  !
  DO i = 1, 60
     DO j = 1, ny1 
        DO k = 1, nz1
            temp_data(i,j,k) = 200.0
        ENDDO
     ENDDO
  ENDDO
  DO i = 61, 120
     DO j = 1, ny1 
        DO k = 1, nz1
            temp_data(i,j,k) = 250.0
        ENDDO
     ENDDO
  ENDDO
  DO i = 121, nx1
     DO j = 1, ny1 
        DO k = 1, nz1
            temp_data(i,j,k) = 200.0
        ENDDO
     ENDDO
  ENDDO
  !
  ! Create a new file using the default properties.
  !
  
  CALL h5fcreate_f(filename, H5F_ACC_TRUNC_F, file, hdferr)

  ! PART 1: temperature

  !
  ! Create dataspace.  Setting size to be the current size.
  !
  CALL h5screate_simple_f(3, temp_dims, space, hdferr)
  !
  ! Create the dataset.  We will use all default properties for this
  ! example.
  !
  CALL h5dcreate_f(file, RADIATION, h5t_ieee_f32le, space, dset, hdferr)
  !
  ! Write the data to the dataset.
  !
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, temp_data, temp_dims, hdferr)
  !
  !
  ! add 1st attribute: units
  !
  dimsf = 1
  attr_data= "Watts/(m^2)"
  strlen= 11
  !
  ! Create scalar data space for the attribute.
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  ! Create datatype for the attribute.
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  !Create dataset attribute.
  !
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, hdferr)
  !
  ! Write the attribute data.
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  !
  ! Close the dataspace and attribute.
  !
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 2nd attribute: fillvalue
  !
  dimsf = 1
  fill_data(1)= -999.0
  CALL h5screate_simple_f(1, dimsf, space, hdferr)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, hdferr)
  CALL h5acreate_f(dset, FILLVALUE, atype_id, space, &
                      attr_id, hdferr)
  dimsf = 1
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, fill_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 3rd attribute: long name
  !
  dimsf = 1
  attr_data= "outgoing long-wave radiation"
  strlen= 28
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, hdferr)
  dimsf = 3
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 4th attribute: coordinates
  !
  dimsf = 3
  coorlist(1)= "latitude"
  coorlist(2)= "longitude"
  coorlist(3)= "pressure"
  strlen= 10
  CALL h5screate_simple_f(1, dimsf, space, hdferr)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  CALL h5acreate_f(dset, COORDINATES, atype_id, space, &
                      attr_id, hdferr)
  dimsf = 1
  CALL h5awrite_f(attr_id, atype_id, coorlist, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)


  CALL h5dclose_f(dset , hdferr)



  ! PART 2: latitude

  ! initialize data
  lat_data(1)= -90.0
  DO i = 2, nx1
    lat_data(i) = lat_data(i-1)+1.0
  ENDDO

  !
  CALL h5screate_simple_f(1, lat_dims, space, hdferr)
  !
  CALL h5dcreate_f(file, LATITUDE, h5t_ieee_f32le, space, dset, hdferr)
  !
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, lat_data, lat_dims, hdferr)
  !
  ! add 1st attribute: long_name
  !
  dimsf = 1
  attr_data= "latitude"
  strlen= 8
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 2nd attribute: units
  !
  dimsf = 1
  attr_data= "degrees_north"
  strlen= 13
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)


  CALL h5dclose_f(dset , hdferr)



  ! PART 3: longitude

  ! initialize data
  lon_data(1)= -180.0
  DO i = 2, ny1
    lon_data(i) = lon_data(i-1)+1.0
  ENDDO

  !
  CALL h5screate_simple_f(1, lon_dims, space, hdferr)
  !
  CALL h5dcreate_f(file, LONGITUDE, h5t_ieee_f32le, space, dset, hdferr)
  !
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, lon_data, lon_dims, hdferr)

  !
  ! add 1st attribute: long_name
  !
  dimsf = 1
  attr_data= "longitude"
  strlen= 9
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 2nd attribute: units
  !
  dimsf = 1
  attr_data= "degrees_east"
  strlen= 12
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)


  CALL h5dclose_f(dset , hdferr)





  ! PART 4: pressure

  ! initialize data
  press_data(1)= 1000.0
  DO i = 2, nz1
    press_data(i) = press_data(i-1)-37.5
  ENDDO

  !
  CALL h5screate_simple_f(1, press_dims, space, hdferr)
  !
  CALL h5dcreate_f(file, PRESSURE, h5t_ieee_f32le, space, dset, hdferr)
  !
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, press_data, press_dims, hdferr)

  !
  ! add 1st attribute: long_name
  !
  dimsf = 1
  attr_data= "pressure"
  strlen= 8
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  !
  ! add 2nd attribute: units
  !
  dimsf = 1
  attr_data= "hpa"
  strlen= 3
  !
  CALL h5screate_f(H5S_SCALAR_F, space, hdferr)
  !
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, hdferr)
  CALL h5tset_size_f(atype_id, strlen, hdferr)
  !
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, hdferr)
  !
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, hdferr)
  !
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, attr_data, dimsf, hdferr)
  CALL h5sclose_f(space, hdferr)
  CALL h5aclose_f(attr_id, hdferr)

  CALL h5dclose_f(dset , hdferr)

  !
  ! Close and release resources.
  !
  CALL h5fclose_f(file , hdferr)
  !
END PROGRAM main

File: hdf5/c/hdf5_3D.c
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 * Copyright by The HDF Group.                                                   !
 * All rights reserved.                                                          !
 *                                                                               !
 * The full HDF5 copyright notice, including terms governing use, modification,  !
 * and redistribution, is contained in the file COPYING.  COPYING can be found   !
 * at the root of the source code distribution tree.                             !
 * For questions contact the help desk at help@hdfgroup.org                      !
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

/*
 * created by:  Isayah Reed <ireed@hdfgroup.org>
 *              May  4, 2011
 *
 * Purpose:
 *  This is 1 of 4 simple HDF5 programs that demonstrates how to add CF attributes
 *  to an h5 file. This example shows how to use 3-D datasets. A file is created 
 *  with 4 datasets: radiation, latitude, longitude, pressure. Radiation is a 3-D
 *  dataset that contains the CF attributes: units, fillvalue, long_name, and 
 *  coordinates. Latitude, longitude, and pressure contain the CF attributes units 
 *  and long_name. Data is written to 3D.h5
 *
 * Compile:
 *  gcc hdf5_3D.c -I$(HDF5_DIR)/include -L$(HDF5_DIR)/lib -lhdf5_hl -lhdf5 -lz -lm
 *
*/


#include "hdf5.h"
#include "string.h"
#include <stdlib.h>

#define H5FILE_NAME "3D.h5"
#define RADIATION "Radiation"
#define LATITUDE "latitude"
#define LONGITUDE "longitude"
#define PRESSURE "pressure"


int main (void)
{
    hid_t       file, dataset, att;         /* file, dataset, attribute handles */
    hid_t       floatType, stringType, arrayType; /* datatypes */
    hid_t       floatSpace, stringSpace, arraySpace; /* dataspaces */
    hsize_t     dimsa[2], dimsa3[3], dimsf[1], dimsc[2];  /* dataset dimensions */
    herr_t      status;
    int         i, j, k;
    char *longname, *units;
    char *coorlist[3]= {"latitude","longitude","pressure"};
    float *rad_array, lat_array[180],
            lon_array[360], press_array[20], fillvalue;


   file = H5Fcreate(H5FILE_NAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

/* PART 1: radiation */
  
    /* allocate the radiation array */
    /* rad_array[180][360][120] */
    rad_array = (float *) malloc(180*360*120*sizeof(float));
/*    rad_array= (float***)malloc(180*sizeof(float**));
    for(i=0; i<180; i++)
    {
        rad_array[i]= malloc(360*sizeof(float*));
        for(j=0; j<360; j++)
            rad_array[i][j]= malloc(120*sizeof(float));
    }

*/
    /* initialize array */
    for(i=0; i<180*360*20; i++)
           	*(rad_array+i)= 280.0;


   /* initialize handles */
    floatType= H5Tcopy(H5T_NATIVE_FLOAT);
    H5Tset_precision(floatType, 32);
    stringType= H5Tcopy(H5T_C_S1);
    stringSpace= H5Screate(H5S_SCALAR);

    dimsa3[0]= 180;
    dimsa3[1]= 360;
    dimsa3[2]= 20;
    floatSpace= H5Screate_simple(3, dimsa3, NULL);
    dataset= H5Dcreate(file, RADIATION, floatType, floatSpace,
    	H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
    	H5P_DEFAULT, rad_array);
			
   
   
   /* add 1st attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("Watts/(m^2)"));
    att= H5Acreate(dataset, "units", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"Watts/(m^2)");
    H5Aclose(att);
   
    /* add 2nd attribute */
    dimsf[0]= 1;
    fillvalue= -999.0;
    floatSpace= H5Screate(H5S_SCALAR);
    att= H5Acreate(dataset, "_FillValue", floatType, floatSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,floatType,&fillvalue);
    H5Aclose(att);
   
    /* add 3rd attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("outgoing long-wave radiation"));
    att= H5Acreate(dataset, "long_name", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"outgoing long-wave radiation");
    H5Aclose(att);


   
   /* add 4th attribute */
    dimsa[0]= 3;
    arraySpace= H5Screate_simple(1, &dimsa[0], NULL);
    arrayType= H5Tcopy(H5T_C_S1);
    status= H5Tset_size(arrayType, H5T_VARIABLE);
    att= H5Acreate(dataset, "coordinates", arrayType, arraySpace, H5P_DEFAULT,
		H5P_DEFAULT);
    status= H5Awrite(att,arrayType,coorlist);
    H5Aclose(att);


    H5Dclose(dataset);
   




/* PART 2: latitude  */

    /*initialize: arithmetic sequence from -90.0 - 89.0 */
    lat_array[0]= -90.0;
    for(i = 1; i < 180; i++)
        lat_array[i] = lat_array[i-1]+1.0;

    dimsf[0] = 180;
    floatSpace = H5Screate_simple(1, dimsf, NULL);


    dataset = H5Dcreate(file, LATITUDE, floatType, floatSpace,
        H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
        H5P_DEFAULT, lat_array);

    /* add 1st attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("latitude"));
    att= H5Acreate(dataset, "long_name", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"latitude");
    H5Aclose(att);

    /* add 2nd attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("degrees_north"));
    att= H5Acreate(dataset, "units", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"degrees_north");
    H5Aclose(att);

    
   
   
/* Part 3: lon  */

    /*initialize: arithmetic sequence from -180.0 - 179.0 */
   lon_array[0]= -180.0;
   for(i = 1; i < 360; i++)
       lon_array[i] = lon_array[i-1]+1.0;

    dimsf[0] = 360;
    floatSpace = H5Screate_simple(1, dimsf, NULL);

    dataset = H5Dcreate(file, LONGITUDE, floatType, floatSpace,
            H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
         H5P_DEFAULT, lon_array);

    /* add 1st attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("longitude"));
    att= H5Acreate(dataset, "long_name", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"longitude");

    /* add 2nd attribute */
    status= H5Tset_size(stringType, (hsize_t)strlen("degrees_east"));
    att= H5Acreate(dataset, "units", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"degrees_east");


	
/* Part 4: pressure  */


    /* initialize: arithmetic sequence from 1000 to 250 */
    press_array[0]= 1000.0;
    for(i=1; i<20; i++)
	press_array[i]= press_array[i-1]-37.5;
			
    dimsf[0] = 20;
    floatSpace = H5Screate_simple(1, dimsf, NULL);

    dataset = H5Dcreate(file, PRESSURE, floatType, floatSpace,
	    H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
	    H5P_DEFAULT, press_array);
			 
    /* add 1st attribute */
    status= H5Tset_size(stringType, strlen("pressure"));
    att= H5Acreate(dataset, "long_name", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"pressure");

    /* add 2nd attribute */
    status= H5Tset_size(stringType, strlen("hpa"));
    att= H5Acreate(dataset, "units", stringType, stringSpace, H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(att,stringType,"hpa");
		
	
    

    /* Close/release resources. */
/*    for(i=0; i<180; i++)
        for(j=0; j<360; j++)
            free(rad_array[i][j]);
    for(i=0; i<180; i++)
            free(rad_array[i]);
*/

     free(rad_array);
    H5Sclose(floatSpace);
    H5Sclose(stringSpace);
    H5Sclose(arraySpace);
    H5Tclose(floatType);
    H5Tclose(stringType);
    H5Tclose(arrayType);
    H5Fclose(file);

    return 0;
}

